pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -363,71 +365,126 @@ func (s *Service) getAssociatedHost(ctx context.Context) (*infrav1.HetznerBareMe
 
+	mapOfSkipReasons := make(map[string]int)
+
+	availableDevices := 0
 	for i, host := range hosts.Items {
 		if host.Spec.ConsumerRef != nil && consumerRefMatches(host.Spec.ConsumerRef, s.scope.BareMetalMachine) {
 			helper, err := patch.NewHelper(&hosts.Items[i], s.scope.Client)
 			if err != nil {
-				return nil, nil, fmt.Errorf("failed to create patch helper: %w", err)
+				return nil, nil, "", fmt.Errorf("failed to create patch helper: %w", err)
 			}
-			return &hosts.Items[i], helper, nil
+			return &hosts.Items[i], helper, "", nil
 		}
 		if host.Spec.ConsumerRef != nil {
 			continue
 		}
-		if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+		if host.Spec.Status.ProvisioningState != infrav1.StateNone {
 			continue
 		}
+
+		// from now on each "continue" should add an entry
+		// to mapOfSkipReasons.
+		availableDevices++
+
+		// This comes first, because we should not look too deep into machines
+		// which are not in our scope.
+		if !labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
+			mapOfSkipReasons["label-selector-does-not-match"]++
+			continue
+		}
+
 		if host.GetDeletionTimestamp() != nil {
+			mapOfSkipReasons["hbmh-has-deletion-timestamp"]++
+			continue
+		}
+		if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+			mapOfSkipReasons["hbmh-in-maintenance-mode"]++
 			continue
 		}
 		if host.Spec.Status.ErrorMessage != "" {
+			mapOfSkipReasons["hbmh-has-error-message-in-status"]++
 			continue
 		}
 
-		if !labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
-			continue
+		// This comes last, because now we would choose the machine, but we check
+		// if the config is correct.
+		if host.Spec.RootDeviceHints != nil &&
+			host.Spec.RootDeviceHints.IsValid() {
+			// If there are no RootDeviceHints, then the it should be ok that the hbmh gets picked.
+			// Then during registring, the HardwareDetails get fetched. After that the process fails.
+			// This has the benefit, that the user can easily copy+paste the WWNs from the HardwareDetails
+			// to the RootDeviceHints.
+			// Doing that without first choosing the hbmh would be nice, there is a feature request:
+			// https://github.com/syself/cluster-api-provider-hetzner/issues/1166
+			if s.scope.BareMetalMachine.Spec.InstallImage.Swraid == 1 {
+				// Machine should have RAID. Skip machines which have less than two WWNs
+				len_wwns := len(host.Spec.RootDeviceHints.Raid.WWN)
+				if len_wwns < 2 {
+					mapOfSkipReasons[fmt.Sprintf("machine-should-use-swraid-but-only-%d-RAID-WWN-in-hbmh", len_wwns)]++
+					continue
+				}
+			} else {
+				// Machine should have no RAID.
+				if host.Spec.RootDeviceHints.WWN == "" {
+					mapOfSkipReasons["machine-should-use-no-swraid-and-no-non-raid-WWN-in-hbmh"]++
+					continue
+				}
+			}
 		}
 
-		if host.Spec.Status.ProvisioningState != infrav1.StateNone {
+		usableHosts = append(usableHosts, &hosts.Items[i])
+	}
+
+	reasons := make([]string, 0, len(mapOfSkipReasons))
```

Comment:
can you put this in a utility function that outputs a string ("reasons")?

Reply:
yes, makes sense. done in ab75bb93